core/src/block_sync.rs

```diff
@@ -25,66 +32,103 @@ pub struct BlockSynchronizer {
-                .await;
-        }
-    }
-}
-
-#[async_trait::async_trait]
-impl Handler<message::Message> for BlockSynchronizer {
-    type Result = ();
-    async fn handle(&mut self, message: message::Message) {
-        message.handle_message(self).await;
-    }
+    p2p: Arc<P2PSystem>,
 }
 
 impl BlockSynchronizer {
-    /// Sends request for latest blocks to a chosen peer
-    async fn request_latest_blocks_from_peer(&mut self, peer_id: PeerId) {
-        message::Message::GetBlocksAfter(message::GetBlocksAfter::new(
-            self.sumeragi.latest_block_hash(),
-            self.peer_id.clone(),
-        ))
-        .send_to(self.broker.clone(), peer_id)
-        .await;
-    }
-
     /// Create [`Self`] from [`Configuration`]
     pub fn from_configuration(
         config: &Configuration,
         sumeragi: Arc<Sumeragi>,
+        p2p: Arc<P2PSystem>,
         peer_id: PeerId,
-        broker: Broker,
-    ) -> Self {
-        Self {
+    ) -> Arc<Self> {
+        Arc::new(Self {
             peer_id,
             sumeragi,
             gossip_period: Duration::from_millis(config.gossip_period_ms),
             block_batch_size: config.block_batch_size,
-            broker,
-            actor_channel_capacity: config.actor_channel_capacity,
+            p2p,
+        })
+    }
+}
+
+/// Initiate the read thread with the main loop `block_sync_read_loop`
+///
+/// # Panics
+///
+/// If thread wasn't spawned
+#[allow(clippy::expect_used)]
+pub fn start_read_loop(block_sync: Arc<BlockSynchronizer>) -> ThreadHandler {
+    // Oneshot channel to allow forcefully stopping the thread.
+    let (shutdown_sender, shutdown_receiver) = tokio::sync::oneshot::channel();
+
+    let thread_handle = std::thread::Builder::new()
+        .name("Block Synchronizer Thread".to_owned())
+        .spawn(move || {
+            block_sync_read_loop(&block_sync, shutdown_receiver);
+        })
+        .expect("Failed to spawn read loop thread handle. You might be hitting up against `ulimits`. Consider increasing the RAM");
+
+    let shutdown = move || {
+        let _result = shutdown_sender.send(());
+    };
+
+    ThreadHandler::new(Box::new(shutdown), thread_handle)
+}
+
+fn block_sync_read_loop(
+    block_sync: &BlockSynchronizer,
+    mut shutdown_receiver: tokio::sync::oneshot::Receiver<()>,
+) {
+    let mut last_requested_blocks = Instant::now();
+    let mut request_blocks_peer_index = 0_usize;
+    loop {
+        // We have no obligations to network delivery so we simply exit on shutdown signal.
+        if shutdown_receiver.try_recv().is_ok() {
+            iroha_logger::debug!("P2P thread is being shut down");
+            return;
+        }
+        block_sync
+            .p2p
+            .poll_network_for_packet(NetworkMessageVariant::BlockSync)
+            .map_or_else(
+                || {
+                    std::thread::sleep(Duration::from_millis(10));
+                },
+                |message| match message {
```

Comment:
We could rewrite this idiomatically with an `if let NetworkMessage::BlockSync(message)`